modulus/datapipes/climate/climate.py

```diff
@@ -0,0 +1,790 @@
+
+import json
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from itertools import chain
+
+import h5py
+import netCDF4 as nc
+import numpy as np
+import scipy
+import torch
+
+try:
+    import nvidia.dali as dali
+    import nvidia.dali.plugin.pytorch as dali_pth
+except ImportError:
+    raise ImportError(
+        "DALI dataset requires NVIDIA DALI package to be installed. "
+        + "The package can be installed at:\n"
+        + "https://docs.nvidia.com/deeplearning/dali/user-guide/docs/installation.html"
+    )
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Callable, Iterable, List, Mapping, Tuple, Union
+
+from modulus.datapipes.climate.utils.zenith_angle import cos_zenith_angle
+from modulus.datapipes.climate.utils.invariant import latlon_grid
+from modulus.datapipes.datapipe import Datapipe
+from modulus.datapipes.meta import DatapipeMetaData
+from modulus.launch.logging import PythonLogger
+
+Tensor = torch.Tensor
+
+
+@dataclass
+class MetaData(DatapipeMetaData):
+    name: str = "Climate"
+    # Optimization
+    auto_device: bool = True
+    cuda_graphs: bool = True
+    # Parallel
+    ddp_sharding: bool = True
+
+
+class ClimateDataSourceSpec:
+    """
+    A data source specification for ClimateDatapipe.
+
+    HDF5 files should contain the following variable with the corresponding
+    name:
+    - `fields`: Tensor of shape (num_timesteps, num_channels, height, width),
+      containing climate data. The order of the channels should match the order
+      of the channels in the statistics files. The statistics files should be
+      `.npy` files with the shape (1, num_channels, 1, 1).
+    The names of the variables are found in the metadata file found in
+    `metadata_path`.
+
+    NetCDF4 files should contain a variable of shape
+    (num_timesteps, height, width) for each variable they provide. Only the
+    variables listed in `variables` will be loaded.
+
+    Parameters
+    ----------
+    data_dir : str
+        Directory where climate data is stored
+    name: Union[str, None], optional
+        The name that is used to label datapipe outputs from this source.
+        If None, the datapipe uses the number of the source in sequential order.
+    file_type: str
+        Type of files to read, supported values are "hdf5" (default) and "netcdf4"
+    stats_files: Union[Mapping[str, str], None], optional
+        Numpy files to data statistics for normalization. Supports either a channels
+        format, in which case the dict should contain the keys "mean" and "std", or a
+        named-variable format, in which case the dict should contain the key "norm" .
+        If None, no normalization will be used, by default None
+    metadata_path: Union[Mapping[str, str], None], optional for NetCDF, required for HDF5
+        Path to the metadata JSON file for the dataset (usually called data.json).
+    channels : Union[List[int], None], optional
```

Comment:
You could imagine removing `channels` altogether and using `variables` for both where you give a dict of key and list of channel pairs. For example, the default would be `{'field': [0, 1, 3}`. Then you would basically combine the variables and channels. If None is given it will just take all and act as if you are just giving the name of the variables. What do you think?

Reply:
I'm not totally sure how you intend the dict to work - why would a variable name map to multiple channels? But I do agree it's better to specify the variables by name than by channel number. The channels option is there more for backward compatibility. Maybe we can deal with this later?

Reply:
Yeah, Ill think through this more and see if I have a solution. Lets not worry about it now.